javalanglib/src/main/scala/java/lang/Math.scala

```diff
@@ -305,13 +305,35 @@ object Math {
     if (a >= Integer.MIN_VALUE && a <= Integer.MAX_VALUE) a.toInt
     else throw new ArithmeticException("Integer overflow")
 
+  def floorDiv(a: scala.Int, b: scala.Int): scala.Int = {
+    if (b == 0)
+      throw new ArithmeticException("/ by 0")
+    if ((a < 0) == (b < 0) || a % b == 0) a / b
+    else a / b - 1
+  }
+
+  def floorDiv(a: scala.Long, b: scala.Long): scala.Long = {
+    if ((a < 0) == (b < 0) || a % b == 0) a / b
+    else a / b - 1
```

Comment:
Consider this alternative formulation which trades one division for one multiplication. Multiplications are much faster than divisions, especially for longs:
```scala
val r = a / b
if ((a < 0) == (b < 0) || r * b == a) r
else r - 1
```
For symmetry, the `Int` version should do the same thing, even though the division overhead there is negligible.